### PR TITLE
openstack: Create machine resources programmatically

### PR DESCRIPTION
**Pull request by flaper87**
Create the machine yaml files using go types rather than go template. This makes testing easier, it guarantees support for a specific openstack actuator version, and allows for more complex logic during the generation.

### CODE REVIEW

File: pkg/asset/machines/openstack/machines.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+			Kind:       "OpenstackProviderSpec",
+		},
+		Flavor: mpool.FlavorName,
+		/*RootVolume: openstackprovider.RootVolume{
+			VolumeType: pointer.StringPtr(mpool.Type),
+			Size:       pointer.Int64Ptr(int64(mpool.Size)),
+		},*/
+		Image:          osImage,
+		CloudName:      platform.Cloud,
+		CloudsSecret:   &corev1.SecretReference{Name: cloudsSecret},
```
Comment (tomassedovic) [inline pkg/asset/machines/openstack/machines.go]: This seems to generate:

```yaml
cloudsSecret:
    name: openstack-credentials
```

but the previous machine template had just: `cloudsSecret: "openstack-credentials"`:

https://github.com/openshift/installer/pull/1042/files#diff-50d413300f38115e1c37fec85d8d7865R47

File: pkg/asset/machines/openstack/machines.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	return machines, nil
+}
+
+func provider(clusterID, clusterName string, platform *openstack.Platform, mpool *openstack.MachinePool, osImage string, az string, role, userDataSecret string) (*openstackprovider.OpenstackProviderSpec, error) {
+	return &openstackprovider.OpenstackProviderSpec{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "openstackproviderconfig.k8s.io/v1alpha1",
+			Kind:       "OpenstackProviderSpec",
```
Comment (tomassedovic) [inline pkg/asset/machines/openstack/machines.go]: The `Kind` used to be `OpenStackMachineProviderConfig`, though I suspect `OpenstackProviderSpec` is the correct value now?

File: pkg/asset/machines/openstack/machines.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+		Networks: []openstackprovider.NetworkParam{
+			{
+				Filter: openstackprovider.Filter{
+					Tags: fmt.Sprintf("%s=%s", "openshiftClusterID", clusterID),
+				},
+			},
+		},
+		AvailabilityZone: az,
+		SecurityGroups:   []string{role},
+		// TODO(flaper87): Trunk support missing. Need to add it back
```
Comment (tomassedovic) [inline pkg/asset/machines/openstack/machines.go]: In addition to trunk support, the new templates don't seem to have the region placement set:

https://github.com/openshift/installer/pull/1042/files#diff-50d413300f38115e1c37fec85d8d7865R50

File: pkg/asset/machines/openstack/machines.go (hunk 2 above)
Comment (flaper87) [reply pkg/asset/machines/openstack/machines.go]: correct

File: pkg/asset/machines/openstack/machines.go (hunk 3 above)
Comment (flaper87) [reply pkg/asset/machines/openstack/machines.go]: Will check and add it if needed.

Comment (flaper87) [reply pkg/asset/machines/openstack/machines.go]: yeah, `placement` is not a valid key in the actuator: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/master/pkg/apis/openstackproviderconfig/v1alpha1/types.go#L30

**Comment by tomassedovic** [comment]
I'm seeing the following issue in the `openshift` log on the bootstrap node:

```
$ journalctl -f -u openshift
Jan 28 16:41:53 ostest-bootstrap openshift.sh[3347]: kubectl create --filename ./99_openshift-cluster-api_worker-machineset.yaml failed. Retrying in 5 seconds...
Jan 28 16:41:59 ostest-bootstrap openshift.sh[3347]: The MachineSet "ostest-worker-" is invalid: metadata.name: Invalid value: "ostest-worker-": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

The `/etc/mcc/bootstrap/99_openshift-cluster-api_worker-machineset.yaml` file contains this:

```yaml
apiVersion: v1
items:
- apiVersion: cluster.k8s.io/v1alpha1
  kind: MachineSet
  metadata:
    creationTimestamp: null
    labels:
      sigs.k8s.io/cluster-api-cluster: ostest
      sigs.k8s.io/cluster-api-machine-role: worker
      sigs.k8s.io/cluster-api-machine-type: worker
    name: ostest-worker-
    namespace: openshift-cluster-api
  spec:
    replicas: 3
    selector:
      matchLabels:
        sigs.k8s.io/cluster-api-cluster: ostest
        sigs.k8s.io/cluster-api-machineset: ostest-worker-
    template:
      metadata:
        creationTimestamp: null
        labels:
          sigs.k8s.io/cluster-api-cluster: ostest
          sigs.k8s.io/cluster-api-machine-role: worker
          sigs.k8s.io/cluster-api-machine-type: worker
          sigs.k8s.io/cluster-api-machineset: ostest-worker-
```

I guess the worker names are missing the unique suffix?

**Comment by tomassedovic** [comment]
https://github.com/openshift/cluster-kube-controller-manager-operator/pull/135#issuecomment-458080457

/lgtm

**Pull request merged**